Implement reparenting in tree_arena

not-my-profile opened the pull request:
This commit firstly implements root_ids and child_ids iterators so that the reparenting implementation can be properly tested.

Reparenting is then implemented in both the safe and the unsafe tree. For the unsafe tree no new unsafe code is introduced.

To make the tests more readable a new Node struct is introduced in the test code so that the tree after the reparenting can be compared with a single assert_eq!.

DJMcNab commented on tree_arena/src/tree_arena_safe.rs:
```diff
@@ -188,6 +198,62 @@ impl<T> TreeArena<T> {
         parents_map.get_id_path(id.into(), None)
     }
+
+    // Moves the given child (along with all children) to the new parent.
```
This should probably be a doc comment?

not-my-profile replied:
Oh yes, thanks good catch!

DJMcNab commented:
cc @KGrewal1, would you be willing to give this a review?

DJMcNab commented on tree_arena/src/tree_arena_safe.rs:
```diff
@@ -188,6 +198,62 @@ impl<T> TreeArena<T> {
+
+/// An iterator over the children ids of a [`TreeArena`] in arbitrary order.
+#[derive(Debug)]
+pub struct RootIdIter<'a>(vec::IntoIter<NodeId>, PhantomData<&'a ()>);
```
What's the value of this lifetime parameter? I presume it's to match the unsafe version?
It might be worth adding a single line comment to that effect

not-my-profile replied:
Yes indeed. I've added such a comment :)

KGrewal1 replied:
I think the alternative would be storing `self.roots.keys().copied()` directly in the struct as the ugly type `Copied<hashbrown::hash_map::Keys<'_, u64, tree_arena_safe::TreeNode<T>>>`  and then having PhantomData<T> in the unsafe API - advantage is that this avoids the call to collect which isn't really necessary (as call `.copied().collect().into_iter()`, disadvantage is that it now has a generic type T associated which is not ideal (can't see a way to store a less concrete type)

KGrewal1 replied:
Though as I believe the safe arena is still generally used, maybe is worth caring about the extra allocation in that path?

KGrewal1 replied:
Or does this actually need a concrete type at all and just be a function like:
```
    /// An iterator visiting all root ids in arbitrary order.
    pub fn root_id_iter(&self) -> impl Iterator<Item = NodeId> + use<'_, T> {
        self.roots.keys().copied()
    }
  ```
  
  for the safe version and
  ```
      /// An iterator visiting all root ids in arbitrary order.
    pub fn root_id_iter(&self) -> impl Iterator<Item = NodeId> + use<'_, T> {
        self.roots.iter().copied()
    }
    ```
    
    for the unsafe?

not-my-profile replied:
Oh yes indeed just `-> impl Iterator<Item = NodeId>` suffices ... it doesn't even need `+ use<'_, T>`.

KGrewal1 replied:
I think similar could be done with the `ArenaRef` into iter pattern too:

for safe:

```
 pub fn iter_children(&self) -> impl Iterator<Item = NodeId> {
        self.children.children.keys().copied()
    }
 ```

for unsafe:

```
    pub fn iter_children(&self) -> impl Iterator<Item = NodeId> {
        self.children
            .parent_arena
            .find_node(self.id())
            .into_iter()
            .flat_map((|c: &TreeNode<T>| &c.children))
            .copied()
    }
  ```

It avoids some messy types like `Copied<hashbrown::hash_map::Keys<'a, u64, TreeNode<T>>>` and `FlatMap<option::IntoIter<&'a TreeNode<T>>, &'a Vec<NodeId>, GetChildren<T>>`, takes `&self` instead of `self` and I think maintains least suprisal vs an `IntoIter` impl  on the `ArenaRef` iterating over specifically the `NodeId` - maybe both this and the function `root_ids` could have the same name too for ease of use?

not-my-profile replied:
I don't really mind messy types as long as they're hidden in a type but I agree that IntoIterator having specifically Item=NodeId is a bit unintuitive and an explicit `.child_ids()` also makes the calling code more readable.

Regarding the naming I prefer to name the iterator over the root ids `root_ids` rather than `child_ids` because the latter would imply that the roots have some parent node, which isn't the case.

DJMcNab commented on tree_arena/src/tree_arena_unsafe.rs:
```diff
@@ -276,6 +286,64 @@ impl<T> TreeArena<T> {
         self.data_map.get_id_path(id.into(), None)
     }
+
+    // Moves the given child (along with all children) to the new parent.
```
This should also be a doc comment

DJMcNab commented on tree_arena/src/tree_arena_unsafe.rs:
```diff
@@ -276,6 +286,64 @@ impl<T> TreeArena<T> {
+
+        assert_ne!(
+            child_id, new_parent_id,
+            "expected child to be different from new_parent"
```
```suggestion
            "expected child ({child_id}) to be different from new_parent ({new_parent_id})"
```

As a way to give a bit more context. It doesn't add all that much value, but it shouldn't really hurt either.
This applies to most of the asserts added here

DJMcNab reviewed (approved):
This is looking really good. I'd like to get at least one more set of eyes on this, as I'm not overly familiar with the tree arena code

KGrewal1 commented:
Happy to give this a review, will look at it over this weekend!

KGrewal1 commented:
On the unsafe side:
- split `find_inner` into `find_node` (w/ internal unsafe but same precondition as before) and `find_inner` calling this all seems great
- implemented reparenting logic - see comment re order of steps moving from one parent to the other but otherwise good

On the safe side new reparenting logic seems fine

And then see comments regarding the node iterator interface, but otherwise looks great

not-my-profile commented:
@KGrewal1 Thanks, I addressed your feedback, except:

> see comment re order of steps moving from one parent to the other

I can't find that comment, perhaps you forgot to submit it?

KGrewal1 commented:
It was this comment https://github.com/linebender/xilem/pull/984#discussion_r2103203816

not-my-profile commented:
Huh there appears to be some GitHub bug. While GitHub shows me a (cut off) preview of your comment when I hover over your link, when I click the link it just takes me to this PR page where I cannot see that comment anywhere. Could you please repeat it?

KGrewal1 commented:
It was essentially moving the insertion into the parents hash map in the unsafe version to before the addition of the node to the new parent list instead of after so that it goes 


[old parent, old parent] -> [None, old parent] -> [None, new parent] -> [new parent, new parent]

instead of

[old parent, old parent]->[None, old parent] -> [new parent, old parent] -> [new parent, new parent]

and so avoiding the intermediate inconsistent state of [new parent, old parent]

not-my-profile commented:
I don't really follow. Moving the child takes 3 operations, no matter in which order you do them the state will always be inconsistent until all of them have been done. I think for readability it's nice to firstly update the `child_arr` and then the parent reference.

KGrewal1 commented:
I think in terms of consistency having it in [None, New] is preferable to [New, Old] as cannot access the child from the None parent (but can from the New) - this should only occur in case of an unwind in the middle of the sequence though so don't think this is a blocker

KGrewal1 commented:
I think it's fine and so don't think it's a blocker to merge, but also not to big a change to stop it going to the [new, old] state either (in case that does change ever)

KGrewal1 commented:
LGTM @DJMcNab @not-my-profile !

DJMcNab commented:
Fwiw, something not being `UnwindSafe` is not something we can rely upon for safety. That is, unwind safety isn't talking about Rust's memory safety guarantees, unfortunately. That is, if:
1) it's possible for this code to panick partway through; AND
2) That panic leaves us in an inconsistent state; AND
3) that inconsistent state allows UB for any later operation

Then some remedial action needs to be taken (such as aborting when this code panicks, or something else).
This isn't high enough on my priority list for me to determine if that's the case myself. If @KGrewal1 asserts than any one (ideally any two!) of the conditions above are false, that's good enough for me.

DJMcNab commented:
Given that:
1) We *think* that the first condition doesn't apply
2) We don't actually ever use the unsafe tree arena

I'm happy to land this and accept that improved robustness as a follow-up, unblocking the work of integrating this in Masonry. Thanks @KGrewal1 for your excellent review.
